### PR TITLE
[Mono] [Arm64] Add SIMD support for vector ctors

### LINKED ISSUE #97815: [Mono] [Arm64] Add SIMD support for vector constructors using other vectors
We have added SIMD support for `Vector2` and `Vector3` on ARM64. Add support for following constructors

- [ ] `Vector3 (Vector2, float)`
- [ ] `Vector4 (Vector3, float)`
- [ ] `Vector4 (Vector2, float, float)`

### PR DESCRIPTION
**Pull request by jkurdek**
Should fix #97815

### CODE REVIEW

**Review by vargaz — APPROVED** [review]

File: src/mono/mono/mini/simd-intrinsics.c [hunk 1]
```diff
@@ -2744,37 +2744,36 @@ emit_vector_2_3_4 (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignature *f
...
-		// FIXME: These don't work since Vector2/Vector3 are not handled as SIMD
-#if 0
 		} else if (len == 3 && fsig->param_count == 2 && fsig->params [0]->type == MONO_TYPE_VALUETYPE && fsig->params [1]->type == etype->type) {
 			/* Vector3 (Vector2, float) */
 			int dreg = load_simd_vreg (cfg, cmethod, args [0], NULL);
-			ins = emit_simd_ins (cfg, klass, OP_INSERT_R4, args [1]->dreg, args [2]->dreg);
-			ins->inst_c0 = 2;
+			MonoInst* vec_ins = args [1];
+			if (COMPILE_LLVM (cfg)) {
+				vec_ins = emit_simd_ins (cfg, klass, OP_XWIDEN, args [1]->dreg, -1);
```
Comment (matouskozak) [inline src/mono/mono/mini/simd-intrinsics.c:2752]: LGTM, good job!
I noticed that there are `System.Numerics.Vectors` red on linux bionic arm64 but it looks infrastructure related. Possibly re-run of CI could make them green.

Comment (jkurdek) [reply src/mono/mono/mini/simd-intrinsics.c:2752]: Checked some jobs before and after changes to vector3. `System.Numerics.Vectors` seems to be red on all of them, along with bunch of different tests on linux bionic arm64, I think it is not related.

**Comment by vargaz** [comment]
This seems to cause regressions on wasm:
https://dev.azure.com/dnceng-public/public/_build/results?buildId=556904&view=logs&jobId=b455efc5-4fe2-50ab-5106-412ad73c23a0&j=b455efc5-4fe2-50ab-5106-412ad73c23a0&t=607d8ddd-2371-58ff-95a3-14f9242e8c35

**Comment by vargaz** [comment]
Testcase:
```
using System;
using System.Threading.Tasks;
using System.Numerics;
using System.Runtime.Intrinsics;

public class Test
{
    public static void foo (Vector3 v) {
        var v2 = new Vector4 (v, 1.0f);
    }

    public static void Main () {
        foo (Vector3.Zero);
    }
}
```
The problems seems to be that the code emits a OP_INSERT_R4, but that expects sreg1 to be a simd vreg, but Vector3 right now is only marked as simd on arm64, not wasm.

**Comment by jkurdek** [comment]
Yes, we do not support neither Vector2 nor Vector3 on WASM. I'll create a PR disabling ctors simd on WASM.

**Comment by vargaz** [comment]
Adding a:
```			if (!mini_class_is_simd (cfg, mono_class_from_mono_type_internal (fsig->params [0])))
				// FIXME:
				return NULL;
```
to the Vector2/Vector3 cases should work around the problem.

**Pull request merged**